test/fixtures/azure-fixture/src/main/java/fixture/azure/MockAzureBlobStore.java

```diff
@@ -0,0 +1,487 @@
+
+        public synchronized String acquire(String proposedLeaseId, int leaseTimeSeconds) {
+            expireIfDue();
+            switch (state) {
+                case Available, Expired -> {
+                    final State prevState = state;
+                    state = State.Leased;
+                    leaseId = proposedLeaseId != null ? proposedLeaseId : UUID.randomUUID().toString();
+                    setExpireTimeMillisSinceEpoch(leaseTimeSeconds);
+                    logger.debug(
+                        "Granting lease, prior state={}, leaseId={}, expires={}",
+                        prevState,
+                        leaseId,
+                        Instant.ofEpochMilli(expireTimeMillisSinceEpoch)
+                    );
+                }
+                case Leased -> {
+                    if (leaseId.equals(proposedLeaseId) == false) {
+                        logger.debug("Mismatch on acquire - proposed leaseId: {}, active leaseId: {}", proposedLeaseId, leaseId);
+                        throw new ConflictException(
+                            "LeaseIdMismatchWithLeaseOperation",
+                            "The lease ID specified did not match the lease ID for the blob/container."
+                        );
+                    }
+                    setExpireTimeMillisSinceEpoch(leaseTimeSeconds);
+                }
+            }
+            return leaseId;
+        }
+
+        public synchronized void release(String requestLeaseId) {
+            expireIfDue();
+            switch (state) {
+                case Available -> throw new ConflictException(
+                    "LeaseNotPresentWithLeaseOperation",
+                    "There is currently no lease on the blob/container."
+                );
+                case Leased, Expired -> {
+                    if (leaseId.equals(requestLeaseId) == false) {
+                        logger.debug("Mismatch on release - submitted leaseId: {}, active leaseId: {}", requestLeaseId, this.leaseId);
+                        throw new ConflictException(
+                            "LeaseIdMismatchWithLeaseOperation",
+                            "The lease ID specified did not match the lease ID for the blob/container."
+                        );
+                    }
+                    state = State.Available;
+                    this.leaseId = null;
+                    this.expireTimeMillisSinceEpoch = Long.MAX_VALUE;
+                }
+            }
+        }
+
+        public synchronized void assertCanWrite(String requestLeaseId) {
+            expireIfDue();
+            switch (state) {
+                case Available, Expired -> {
+                    if (requestLeaseId != null) {
+                        throw new PreconditionFailedException(
+                            "LeaseLost",
+                            "A lease ID was specified, but the lease for the blob/container has expired."
+                        );
+                    }
```

Comment:
IIUC, this means if a lease is never acquired for a blob, we still fail the request if a leaseID is specified for an operation. The behaviour makes sense to me and seems safe. Though the [Azure doc](https://learn.microsoft.com/en-us/rest/api/storageservices/lease-blob?tabs=microsoft-entra-id#remarks) says

> When a lease is active, the lease ID must be included in the request for any of the following operations

I wonder what "a lease is active" means. Does it imply that the leaseId check is performed only if a lease is acquired first? Not necessarily suggesting that's what we should do. But it would be great if we can know for sure.

Reply:
There's a table that indicates how all this works [here](https://learn.microsoft.com/en-us/rest/api/storageservices/lease-blob#outcomes-of-use-attempts-on-blobs-by-lease-state). I linked it from the Javadoc.

I just implemented the subset of states we support (available, expired, leased)